### ISSUE TITLE
Bank Transfer Quarantine

### ISSUE THREAD

**Issue by iramiller** [provenance-io/provenance#1046]
## Summary

Create an intermediate list of address pairs which are allowed to send/receive funds directly.  For account addresses that are not on this list send transfers should be held until the recipient account explicitly authorizes their receipt.  Once an authorization is in place future funds sent from this know address would be allowed without delay.

## Problem Definition

One of the risks for an account based chain versus one that uses a UTXO model is that account balances commingle all incoming funds. While this is much more efficient it can be dangerous in cases where an external account might send funds that are tainted from illicit activity (as defined by OFAC).  These tainted funds can be used to extend the scope of risk to the victims account.

A concrete example of this type of attack has recently occurred on Ethereum where the banned "Tornado Cash" construct is used to "dust accounts" with a small amount of ETH sent through this contract.  Since the OFAC has banned this contract and declared that any funds moved through it are also sanctioned, a malicious entity can send funds via this contract to any other address so that it also becomes banned by association.

An updated version of the "send" system used by the bank module that checks an allow list before moving funds directly into an account will prevent these types of "dusting" attacks.  The recipient account should be able to sign for the sent funds to claim them and in doing so establish an "allowed" address record that lets funds move without quarantine from this address in the future.

The primary disadvantage of including this feature is complexity and increased friction in sending funds between accounts.

## Proposal

- Create a new middleware for the bank module that redirects certain transactions to a module escrow account and creates a ledger entry for their distribution.
- Allow funds that are sent to an address with an existing authorization to be directly sent without a quarantine.
- Add a record for the source address of an account that successfully sends funds to a recipient so that future transfers are allowed without quarantine
- Ensure that all IBC transfers are routed through this same bank quarantine feature

____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

**Comment by iramiller** [comment]
An idea I had is that we could allow the OFAC "block" records to be submitted with a special endpoint  ... and creates a gov proposal for the banning of an account at the same time ... this would enforce an immediate block on an account while starting a formal gov process to decide the validatity of the request ... if the block passed vote it could be permanent ... if it failed the temporary hold would be removed ...

**Comment by iramiller** [comment]
Based on further discussion with @SpicyLemon we explored the idea of expired or rollback of sends when the funds are not accepted.  We determined that there is no need to have a cancel or a rollback capability.  This is because when a send is invoked the funds are moved out of the from account's control and it doesn't matter if the destination account wants the funds or not.  If a user does not accept the funds then they would be permanently held by the bank module account.

**Comment by SpicyLemon** [comment]
After some discussion, here's the current plan:

## Quarantine

Create the ability to opt-in to quarantine. Once an account has opted in, all sends to that account are held by the bank module until accepted or declined. An account using quarantine will have the ability to either accept/decline a single specific send from an address or accept/decline this and all future sends from that address.

Funds from quarantined sends will be held by the bank module account. Declined funds will continue to be held by the bank module account, but won't show up when an account queries for quarantined sends pending their approval. It should still be possible to accept funds after declining them (it'll just be harder to look up the needed info).

Side note: If a quarantined account gets stuck (i.e. not enough funds to pay for an accept Tx), a feegrant can be used.

From a Provenance-specific perspective, part of the KYC process will be to opt the account into quarantine. In the explorer, we can have a page that lists the quarantined sends, the sender, and an indicator if they've been KYC'd, with buttons to accept/decline.

Some options considered and rejected:

* Have a params entry to globally activate/deactivate quarantine for all accounts. Once we thought about the account-level opt-in idea, this was dismissed as unneeded and too broad. It also has the problem of making it difficult to initially fund an account.
* Customize the `EndBlock` (or maybe `BeginBlock`) to use the attribute module to auto-accept sends between accounts that have been KYC'd. This is still something that might be done in future expansions, but needs extra thought about the ramifications. This is also something that would be specific to Provenance (and wouldn't be part of any updates to the `x/bank` module). It should probably also be part of its own module.
* Make the accept send message free. This would be a complex and specific change needed in generic areas. It also opens the door to the SPAM/DOS attacks that gas is meant to prevent. Also, the reason behind this thought was to get around new accounts needing the funds to send an accept Tx before they've got any funds. But with the opt-in strategy, that's not as much of a concern.

The weeds:

* The existing `x/bank` interface cannot be changed. We can add new stuff and alter existing functionality, but we can't change existing function inputs/outputs.
* Add a query to look up quarantined sends to an account. This query shouldn't list ones that were previously declined.
* Add a query to look up quarantined sends to an account from another account. This query should any quarantined funds whether they've been declined or not (maybe include that too).
* Add two new Tx messages/endpoints: one for accept and one for decline. They'd both have a field/flag to indicate one or all. If only accepting one, the funds are released. If accepting all, the funds are sent and a record is added to state. If declining one, the funds are marked as declined. If declining all, the funds are marked as declined and a record is added to state.
* Add two new messages/endpoints for opting in and out of quarantine. All that's needed is the account (which must also sign the Tx).
* Quarantined funds should be cumulative by sender. I.e. If A sends B 10 hash, then later sends 15 hash, account B should see 25 hash waiting in quarantine from A (and only needs a single accept for them).
* When an account opts out, process all quarantined + non-declined funds as accepted and remove their quarantine opt-in record. I'm currently unsure about whether to delete all the auto-response records too. I don't picture it being much data though.
* New things to store in state:
  1. Quarantined funds. Key : `<prefix><receiver><sender>`. Value is a `Coins` and whether they've been declined.
  2. Account opted into quarantine. Key: `<prefix><account>`. Value is a null byte. Entry exists = opted in.
  3. Auto-response. Key: `<prefix><receiver><sender>. Value is `0` for auto-decline, `1` for auto-accept.
* When doing a send:
  1. Check if the receiver account uses quarantine. If not, treat it like a normal send.
  2. If so, check for an auto-response entry.
    * If it's auto-accept, treat it like a normal send.
    * If it's auto-decline, add it to any possibly existing quarantined funds and mark them as declined.
    * If there's no auto-response, add the funds to any possibly existing quarantined funds and make sure they're not marked as declined.

**Comment by SpicyLemon** [comment]
Here's another idea that's been brought up.

## Blacklist for Sending

Another desirable feature discussed is an account blacklist controllable through governance proposals. An account on the blacklist should have their funds locked so they cannot be sent, spent, used, transferred, etc.

One tricky part of this is that there is a need for this to be able to happen very quickly. The current thought is to make a governance proposal that has a larger than usual bond requirement. When such a proposal Tx is sent, addresses would be immediately (without waiting for it to pass) added to a pending list. If it passes the vote, the addresses are added to the blacklist and removed from pending. If it fails, the addresses are just removed from the pending list. Pending addresses would be treated as blacklisted.

This proposal should also allow removing addresses, but that can (probably) happen at the usual governance proposals pace.

Care needs to be taken to prevent bypasses of the blacklist using various modules e.g. authz, bank, feegrants, msgfees, etc. or even a combination of modules. Tipping (added in Cosmos-SDK v0.46) should also be considered.

A bit more research is needed here before creating a plan of action. I think we might be able to do something similar to a vesting account where all the funds are locked. The `x/bank` module's `SendKeeper`'s `BlockedAddr` function is a list of addresses that can't _receive_ funds; so it's not usable for this. But its use is probably a good guide for implementing our blacklist.

Having action taken when the gov proposal is received is a tricky thing. Part of the trickiness can be solved by creating an endpoint that makes the pending additions and then submits a gov proposal on behalf of the sender. Basically, an intermediate gov proposal endpoint.

Having action taken when the gov proposal doesn't pass, though, is tougher. Currently, the only way I can think to do that is to have a special message handler for rejected gov props and have the `gov` module's `EndBlocker` look up and call that (if defined) for rejected messages. That's a pretty significant change though.

**Comment by iramiller** [comment]
Instead of creating a call back for the proposal to clear on failure, have the ban expire on a blockheight shortly after the voting period is set to end. In this way the ban is for a fixed duration unless made permanent through the proposal passing. 

We would need to ensure the bond required to ban an address is very significant. If the action was illegitimate then the penalty must be severe. As an example an anonymous user might elect to ban a provenance foundation address in which case that address could not be used to vote on the proposal to ban it... not voting on your own ban proposal seems appropriate but overall the point still stands that otherwise legitimate addresses might be blocked temporarily (or permanently if the community so decides). 

A related rescind ban proposal with a normal gov proposal cost should also be supported.

**Comment by iramiller** [comment]
We don't want to allow immediate unlock (pending votes) as the assets could be moved quickly to circumvent the bans... Seems like a banned address should not be allowed to successfully sign which would prevent most uses. We would need to review delegation flows too though so that grants could not be used to bypass a ban.

**Comment by SpicyLemon** [comment]
I started in on the quarantine portion and hit a slight snag when I realized I needed to account for `MultiSend`/`InputOutputCoins` too. With those, there isn't always just one sender. I changed the keys for the quarantined funds to be `<prefix><receiver><key suffix>`. If there's only one sender, the `<key suffix>` is the sender. Otherwise, the senders are sorted and joined into a single byte slice that is then sha256 summed. All 32 bytes of that are then the `<key suffix>`. I then had to create an index for those so that when someone accepts funds from a single address, it can properly find the entry. That index is `<prefix><receiver><sender>` -> `<list of suffixes>`.

I also had to add lists of unaccepted and accepted senders to the quarantine records. When an accept is processed, all the proper records are looked up, and the sender is moved from unaccepted to accepted. Once the unaccepted list is empty, the funds are released.

Example:
1. Account Q has auto-accept turned on for Sender 2
2. Sender 1 sends 1 hash to account Q.
3. Sender 1 and Sender 2 do a multi-send to account Q for 2 hash.
   Note: The auto-accept from sender 2 here markes sender 2 as having been accepted. But since sender 1 hasn't been accepted, the funds are still quarantined.
   Also note: This creates a different record from the earlier send since it has different parties involved.
4. Sender 1 and Sender 3 do a multi-send to account Q for 5 hash.
5. Q sends an accept for sender 1.
   Effects:
   * The 1 hash from 1 to Q is marked as accepted.
     Since all senders have been accepted for those funds, 1 hash is then transferred to Q's account and the quarantine record is deleted.
   * The 2 hash from 1 and 2 to Q is marked as accepted from 1.
     Because of the auto-accept Q has for 2, 2 was already listed as accepted.
     Since all senders have been accepted for those funds, 2 hash is transferred to Q's account and the record is deleted.
   * The 5 hash from 1 and 3 to Q is marked as accepted from 1.
     Sender 3 has not yet been accepted for those funds though, so nothing is transferred.
6. Q sends an accept for 3 and that 5 hash is finally transferred to Q's account. 

Other things to note:
* Funds will only skip quarantine if ALL senders are auto-accept for the receiver.
* Funds will automatically be marked as declined if ANY of the senders are auto-decline for the receiver.
* Declined quarantine funds don't show up when a receiver queries for all their quarantined funds, but do show up when querying for quarantined funds from a specific sender.
* The accept and decline messages were updated to allow for providing multiple senders. An accept/decline is processed for each sender provided and is applied to all applicable funds.
* For `MultiSend`/`InputOutputCoins`, each receiver is handled individually for quarantine. E.g. if a multisend has two receivers, one of which is quarantined, the funds for the other receiver are transferred as normal and only the funds for the quarantined account are quarantined. All senders are handled together though. That is, if a multisend has two quarantined receivers, all of the senders are included in both resulting quarantine records.

**Comment by SpicyLemon** [comment]
Here's a hackmd I've put together as an initial point of discussion with the SDK and other chains: https://hackmd.io/@dwedul/BJuR4aQDj

**Comment by SpicyLemon** [comment]
I've created a separate ticket for the sanction stuff: #1306.
That way, I can leave this one to focus on the quarantine module.

**Comment by SpicyLemon** [comment]
In the future, hopefully the SDK has adopted the injectable bank restictions PR: https://github.com/cosmos/cosmos-sdk/pull/14224

Once that's in the SDK, it should be pretty easy to refactor these changes as such:
1. Create a new restriction function in the quarantine keeper that matches the required signature. It should check if the to-account is quarantined. If so (and not auto-accepting from the from-address), record the funds as quarantined and return the funds-holder address. If not, just return the to address. This check should be bypassed if the from-address is the funds-holder account.
3. Inject that restriction into the bank keeper instead of injecting the quarantine keeper.
4. Remove the injectable quarantine keeper stuff from the bank module.

At that point, we might even consider moving the whole module out of our SDK fork and into provenance (or maybe to its own repo as an importable library).

**Comment by SpicyLemon** [comment]
The quarantine module has been merged: https://github.com/provenance-io/cosmos-sdk/pull/335
